community/rfcs/24-08-21-GenAIExample-002-Edge_Craft_RAG.md

```diff
@@ -0,0 +1,118 @@
+application to meet both quality and performance requirements. Total Cost of
+Ownership(TCO) and pipeline optimization techniques are the two main reasons
+to block this process.
+
+#### Total Cost of Ownership
+
+The HW requirement of a typical edge use case is a single host with one of the
+following combinations:
+- Intel(R) Core(TM) Ultra Processor
+- Intel(R) Core(TM) Processor + Intel(R) Iris(R) Xe Graphics
+- Intel(R) Core(TM) Processor + Intel(R) Arc(TM) A-Series Graphics
+- Intel(R) Xeon(R) Processor + Intel(R) Arc(TM) A-Series Graphics
+
+The scenarios with these hardware options block the edge users from using large 
+parameter size LLMs on-prem as well as sophisticated RAG pipeline for their data.
+Thus, the RAG pipeline at edge needs to be highly curated for underlying hardwares
+and suitable models accordingly.
+
+#### RAG Pipeline Optimization Techniques
+
+Tuning RAG pipeline is a systematic problem. First, the quality depends on the
+result of each stage in the pipeline as well as the end-to-end outcome. Second,
+optimization could be a trade-off among the metrics. It is difficult to decide one
+answer is better than another if it is slightly not accurate in number but more
+relevant to the query. Third, the optimization techniques may not intuitively
+reflect to metrics improvements. E.g., recrusive retrieval may contribute to
+improving the recall and context relevancy, or may not.
+
+### Motivation
+
+Edge Craft RAG (EC-RAG) is a customizable, tunable and production-ready
+Retrieval-Augmented Generation system for edge solutions. It is designed to
+curate the RAG pipeline to meet hardware requirements at edge with garanteed
+quality and performance.
+
+From quality perspective, EC-RAG is tunable in the indexing, retrieving,
+reranking and generation stages for particular edge use cases. From performance
+perspective, the pipeline is consolidated in a single service to eliminate the
+overhead of inter-service communication on a single host. Meanwhile, the inferencing
+stages like embedding, reranking and generation are optimized for Intel(R) Iris(R)
+Xe Graphics and Intel(R) Arc(TM) A-Series Graphics.
+
+### Design Proposal
+
+EC-RAG is composed of the following components:
+- UI for doc loading and interactive chatbot.
+- Gateway
+- Mega-service with a single micro-services for the tunable* EC-RAG pipeline.
+- LLM serving microservice optimized for Intel(R) Iris(R) Xe Graphics and Intel(R) Arc(TM) A-Series
+Graphics
+- VectorDB microservice optimized for Intel(R) Iris(R) Xe Graphics and/or Intel(R) Arc(TM) A-Series
+- Docker compose file to launch the UI, Mega/Micro-services
+
+> [!NOTE]
+> *Advanced tuning EC-RAG will need a tool co-piloting with the pipeline which will be described in
+> a separate doc
+
+Below diagram illustrates the overall components of EC-RAG:
+![EC-RAG Diagram](Edge_Craft_RAG.png)
+
+The EC-RAG pipeline will expose 3 types of REST API endpoint:
+- **/v1/data** for indexing
```

Comment:
Why define new endpoint for data indexing? The dataprep microservice in OPEA provides the endpoint "/v1/dataprep".

Reply:
Thanks for your comment. The consideration of not using "/v1/dataprep" is as follows:
- We are not using the dataprep microservice, due to 1) EC-RAG pipeline could be curated including different data ingestion approaches. 2) Currently, we are not binding the storage to any specific databases.
- The "/v1/dataprep" is not defined either in _class MegaServiceEndpoint(Enum)_ or in _class MicroServiceEndpoint(Enum)_
- Since EC-RAG is not a standalone microservice for data preparation, using the same endpoint as the dataprep microservices may introduce misunderstanding on not following the next level routes, e.g. we use a GET on "/v1/data/{name}" to get a file with name, instead of using "/v1/dataprep/get_file".